Test real blocks

Sword-Smith opened the pull request:
test: Recover state from real main net blocks
    
Add a test of `GlobalState::bootstrap_from_directory` using real data
from main net. The test uses the `blk0.dat` file from a node that has
been online since genesis, thus containing reorganizations that some
previously untested branches of `GlobalState::bootstrap_from_directory`.
    
This test also serves as a check that the first 113 main net blocks are
still considered valid by future versions of the software.

Closes #515 

This PR introduces a big file, `blk0.dat`, added with `git lfs`. But an alternative would be to download the `blk0.dat` from our test servers instead, and store it in `test_data` so that it would only ever have to be downloaded once.

dan-da commented:
> But an alternative would be to download the blk0.dat from our test servers instead, and store it in test_data so that it would only ever have to be downloaded once.

This seems preferable to me because:
1. It keeps the repo small(er) for anyone that doesn't need to run unit tests.
2. It still works for anyone that does not have git-lfs installed, but does wish to run unit tests.
3. in general seems to be more out-of-the-box compatible with git distros on other OSes, etc.

I note that git-lfs is installed by default on ubuntu 24.04, but according to gemini LLM:
* windows: git lfs present
* macos: requires install by homebrew
* BSDs: not installed by default
* other linux distros(debian, arch, fedora): not installed by default

and even if installed by default, ti requires user to `git lfs install`.

dan-da reviewed (changes requested):
I'd like to switch to downloading the blk0.dat within the unit test if we can, for reasons outlined in a previous comment.

just for reference:  I performed a timed git checkout of this branch via my slow connection.  It took 8 minutes, 48 seconds.

I'm not sure the GlobalStateLock constructor change is needed and/or I think it can be done more cleanly.  see inline comments.

It's likely worth inviting another reviewer for another set of eyes.

Sword-Smith commented on src/models/state/mod.rs:
```diff
@@ -183,6 +187,62 @@ impl GlobalStateLock {
+        genesis: Block,
+        cli_args: cli_args::Args,
+        rpc_server_to_main_tx: tokio::sync::mpsc::Sender<RPCServerToMain>,
+    ) -> Result<Self> {
```
> I read the commit comment, but I'm still unclear on the motivation for the change. Is this necessary for the PR to "test real blocks"?

Yes. The genesis block under the test flag is different from the real genesis block. Because under the test flag initial difficulty is 6.000 and otherwise it's 1.000.000.000. So the refactor is needed to allow the genesis block to be injected by the caller when creating a new `GlobalState`/`GlobalStateLock`.

Sword-Smith replied:
> If it's just for test purposes, couldn't it be done after GlobalStateLock is created, eg?
These state updates are quite tricky, as the wallet state also needs to be updated with the right block, in which potential premine rewards are received. The genesis block needs to be injected like this.

Sword-Smith commented:
I addressed some concerns:
- Blocks are now downloaded from "proof" servers, cf. 385479c5a20a25604378ba3506278e25b1db1f52
- Less database initialization/opening in `GlobalstateLock`'s constructor, as this has been moved to `GlobalState`'s constructor.

dan-da commented on src/tests/shared.rs:
```diff
@@ -456,6 +452,160 @@ impl<Item> stream::Stream for Mock<Item> {
     }
 }
 
+/// Return path for the directory containing test data, like proofs and block
```
this file is growing large.  maybe it's time to consider splitting into separate file(s).

dan-da commented on src/tests/shared.rs:
```diff
@@ -1330,3 +1458,16 @@ where
     Ok(())
 }
+
+#[cfg(test)]
```
redundant.   This entire file is already #[cfg(test)].

dan-da reviewed (approved):
The `cli_args::Args::default_with_network(network)` seems a nice shortcut.

I left a few more comments/suggestions, but overall it seems ok to merge.

Sword-Smith commented on src/models/state/mod.rs:
```diff
@@ -161,21 +166,17 @@ pub struct GlobalStateLock {
-        net: NetworkingState,
-        cli: cli_args::Args,
-        mempool: Mempool,
+    pub(crate) fn from_global_state(
```
I'll make the constructors for `GlobalState` and `GlobalStateLock` public again.
